common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use libra_logger::prelude::*;
+use std::{collections::HashMap, fmt::Debug, hash::Hash, time::Instant};
+use tokio::time::Duration;
+
+/// Interface for a keyed rate limiter.  This should allow us to
+/// drop in a different keyed rate limiter if required.
+pub trait KeyedRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    /// Determine if a throttle is needed on the key for the given count
+    /// This does not block, or wait and returns `true` if throttled
+    fn throttle(&mut self, key: Key, count: u64) -> bool;
+
+    /// Fills the throttling mechanism for a number of ticks.  Can be manually called
+    fn fill_throttles(&mut self, intervals: u64);
+}
+
+impl<Key: Eq + Hash + Clone + Debug> KeyedRateLimiter<Key> for TokenBucketRateLimiter<Key> {
+    /// Processes a throttle check, for synchronous bookkeeping it also ensures
+    /// all buckets are filled in a timely manner, and buckets exist.
+    /// TODO: Do we want to add a waiting mechanism, this only returns immediately
+    fn throttle(&mut self, key: Key, count: u64) -> bool {
+        // We could be more strict on 0, but there's nothing to throttle then, exit quickly
+        if count == 0 {
+            return false;
+        }
+
+        // Refill buckets if we need to
+        // TODO: Should we do this separately, or continue doing synchronously here
+        self.check_refill(self.last_refresh_time.elapsed());
+
+        // If there's no bucket, let's create a default one
+        self.ensure_bucket(
+            key.clone(),
+            self.default_bucket_size,
+            self.default_fill_rate,
+        );
```

Comment:
HashMap `.entry` API works well for this : )

```rust
let bucket = self.buckets.entry(&key).or_insert_with(|| Bucket::new(bucket_size, fill_rate));
```

Reply:
Nice, my rust was rusty from having time off.